chainstate/src/detail/gen_block_index.rs

```diff
@@ -0,0 +1,57 @@
+use chainstate_types::block_index::BlockIndex;
+use common::chain::block::{consensus_data::BlockRewardTransactable, timestamp::BlockTimestamp};
+use common::chain::{ChainConfig, GenBlock};
+use common::primitives::{BlockHeight, Id};
+use common::Uint256;
+
+/// Generalized block index
+#[derive(Clone, Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum GenBlockIndex<'a> {
+    Block(BlockIndex),
+    Genesis(&'a ChainConfig),
```

Comment:
Why not include the genesis block object directly, instead of complicating it with a reference to ChainConfig?

Reply:
* `ChainConfig` has a pre-calculated genesis ID, so when you use `GenBlockIndex::block_id`, it does not have to serialize and hash it.
* It's a reference just to avoid copying. Works out quite

I'm sure this could be improved, suggestions welcome. The question is how much of this should be done now and how much should be a follow up since this is now blocking other changes.

Reply:
It could just store a reference to genesis and the pre-calculated ID

Reply:
Quite frankly, if we store the whole genesis block in it, that'd be fine. Or the id alone. But you're right, this doesn't matter currently. I was just wondering why this looks like this. Feel free to skip it.

Reply:
> Quite frankly, if we store the whole genesis block in it, that'd be fine. Or the id alone. But you're right, this doesn't matter currently. I was just wondering why this looks like this. Feel free to skip it.

Genesis block contains the initial UTXO set, which may get quite massive, especially with all the pre-mine payouts :smile:

Reply:
Having said that you can skip this, premine in genesis will be very simple. It's gonna be one output to one address. Then the distribution is a whole other problem that can be done later. No need to complicate genesis block.

Reply:
Can this be resolved for the current PR?

Reply:
Just slap TODOs around if you feel the change is big. No need to do it right now.